libraries/AP_UAVCAN_V1/AP_UAVCAN_V1_esc.cpp

```diff
@@ -0,0 +1,353 @@
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Author: Dmitry Ponomarev
+ */
+
+#include <AP_UAVCAN_V1/AP_UAVCAN_V1_esc.h>
+
+#if HAL_ENABLE_LIBUAVCAN_V1_DRIVERS
+
+#include <GCS_MAVLink/GCS.h>
+#include <GCS_MAVLink/GCS_MAVLink.h>
+
+
+extern const AP_HAL::HAL& hal;
+
+
+bool UavcanEscController::init(UavcanSubscriberManager &sub_manager,
+                               UavcanPublisherManager &pub_manager,
+                               CanardInstance &ins,
+                               CanardTxQueue& tx_queue)
+{
+    _pub_setpoint = new UavcanSetpointPublisher(ins, tx_queue, _registers.getPortIdByIndex(UAVCAN_PUB_SETPOINT_ID));
+    if (!pub_manager.add_publisher(_pub_setpoint)) {
+        return false;
+    }
+
+    _pub_readiness = new UavcanReadinessPublisher(ins, tx_queue, _registers.getPortIdByIndex(UAVCAN_PUB_READINESS_ID));
+    if (!pub_manager.add_publisher(_pub_readiness)) {
+        return false;
+    }
+
+    const uint8_t power_register_indexes[4] = {
+        UAVCAN_SUB_POWER_0_ID,
+        UAVCAN_SUB_POWER_1_ID,
+        UAVCAN_SUB_POWER_2_ID,
+        UAVCAN_SUB_POWER_3_ID,
+    };
+
+    const uint8_t dynamics_register_indexes[4] = {
+        UAVCAN_SUB_DYNAMICS_0_ID,
+        UAVCAN_SUB_DYNAMICS_1_ID,
+        UAVCAN_SUB_DYNAMICS_2_ID,
+        UAVCAN_SUB_DYNAMICS_3_ID,
+    };
+
+    const uint8_t status_register_indexes[4] = {
+        UAVCAN_SUB_STATUS_0_ID,
+        UAVCAN_SUB_STATUS_1_ID,
+        UAVCAN_SUB_STATUS_2_ID,
+        UAVCAN_SUB_STATUS_3_ID,
+    };
+
+    int16_t port_id;
+    for (auto esc_idx = 0; esc_idx < 4; esc_idx++) {
+        port_id = _registers.getPortIdByIndex(power_register_indexes[esc_idx]);
+        _sub_power[esc_idx] = new UavcanElectricityPowerTsSubscriber(ins, tx_queue, port_id);
+        if (!sub_manager.add_subscriber(_sub_power[esc_idx])) {
+            return false;
+        }
+
+        port_id = _registers.getPortIdByIndex(dynamics_register_indexes[esc_idx]);
+        _sub_dynamics[esc_idx] = new UavcanDynamicsSubscriber(ins, tx_queue, port_id);
+        if (!sub_manager.add_subscriber(_sub_dynamics[esc_idx])) {
+            return false;
+        }
+
+        port_id = _registers.getPortIdByIndex(status_register_indexes[esc_idx]);
+        _sub_status[esc_idx] = new UavcanStatusSubscriber(ins, tx_queue, port_id);
+        if (!sub_manager.add_subscriber(_sub_status[esc_idx])) {
+            return false;
+        }
+    }
+
+    _is_inited = true;
+    return true;
+}
+
+
+void UavcanEscController::SRV_push_servos()
+{
+    if (!_is_inited) {
+        if (AP_HAL::millis() > last_log_ts_ms + 5000) {
+            last_log_ts_ms = AP_HAL::millis();
+            GCS_SEND_TEXT(MAV_SEVERITY_WARNING, "v1: esc is not inited.");
+        }
+        return;
+    }
+
+    WITH_SEMAPHORE(SRV_sem);
+
+    uint8_t src_idx = 0;
```

Comment:
generally speaking, variable should be created on the scope of their usage.
here src_idx is only used on the for loop, so it should belong to the for loop only.

Reply:
fixed